### PR TITLE
feature/#1858-add-s3-object-data-node

### LINKED ISSUE #1858: Allow S3ObjectDatanodes to use all parameters exposed by AWS APIs
### Description

Today, an s3 object data node can only use a limited number of parameters. 
Taipy should accept all possible parameters for configuring the [boto3 client](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.client), for reading the data with the [get_object](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/get_object.html#get-object) method, and for writing the data with the [put_object](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/put_object.html#put-object) method. 

### Solution Proposed

**DatanodeConfig:** 
The parameters should be passed by the user through the `DataNodeConfig` method
`taipy.core.config.data_node_config._configure_s3_object`.
The goal is to "keep simple things simple", in particular for most common usages.
So, the purpose is not to expose all the parameters in the configure method, but only the main ones. The others should be passed as optional parameters in kwargs properties.

**Datanode:**
All the parameters (usedfor the client constructor, the get_object, and the put_object methods ) should be used.
 
### Acceptance Criteria

- [ ] Ensure the new code is unit tested, and check that the code coverage is at least 90%.
- [ ] Create related issues in taipy-doc for documentation and Release Notes.

### Code of Conduct

- [x] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by AdeshGhadage**
This pull request addresses issue #1858.

Hello Maintainers,

Could you please review my code? I would appreciate any suggestions or feedback you might have.

Thank you!

### CODE REVIEW

File: taipy/core/data/aws_s3.py
Comment (trgiangdo) [inline taipy/core/data/aws_s3.py]: Why are these properties removed?

FYI, these are properties that are required for the DataNode entity

File: taipy/core/data/aws_s3.py
Comment (trgiangdo) [inline taipy/core/data/aws_s3.py]: The arguments for [`boto3.client()`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.client) and [`S3.Client.get_object()`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/get_object.html) and [`S3.Client.put_object()`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/put_object.html) are all different.

Here, using the same parameters for all methods will not work.

We can:
- have a different set of parameters for each method, like `__AWS_S3_CLIENT_PARAMETERS`, `__AWS_S3_GET_OBJECT_PARAMETERS`, and `__AWS_S3_PUT_OBJECT_PARAMETERS`
- take a look into the list of parameters for each method, decide which parameters are the most important, and only expose those.

I would suggest the former approach, since it's the most generic.

**Review by trgiangdo — CHANGES_REQUESTED** [review]
Hello @AdeshGhadage, thank you for your contribution.

The PR has well-handled the `__AWS_REGION` parameter.

And for the `__AWS_S3_OBJECT_PARAMETERS`, it is more complex since we need a more generic solution, which requires updating the configuration as well.

Please take a look at my comment and let me know what you think.

File: taipy/core/data/aws_s3.py
Comment (trgiangdo) [inline taipy/core/data/aws_s3.py]: We can remove the `__AWS_S3_OBJECT_PARAMETERS` here and elsewhere.

**Review by trgiangdo — COMMENTED** [review]
Good progress.

The next step would be:

- updating the configuration for the S3ObjectDataNode [here](https://github.com/Avaiga/taipy/blob/ec5ec7413320186213f748df37153a4a025056f2/taipy/core/config/data_node_config.py#L1076)
- updating the docstring for the `_configure_s3_object()` method with new parameters.
- updating the docstring of the `S3ObjectDataNode` class. It should be very similar to the `_configure_s3_object()` method.

**Comment by AdeshGhadage** [comment]
@trgiangdo @jrobinAV Any other changes ? or some mistake I have made?

File: taipy/core/config/data_node_config.py
Comment (trgiangdo) [inline taipy/core/config/data_node_config.py]:
```suggestion
```

This needs to be removed as well I believe

File: taipy/core/config/data_node_config.py
Comment (trgiangdo) [inline taipy/core/config/data_node_config.py]:
```suggestion

        return cls.__configure(id, DataNodeConfig._STORAGE_TYPE_VALUE_S3_OBJECT, scope, validity_period, **properties)
```

There should not be a tab here

**Review by trgiangdo — COMMENTED** [review]
Some final comments. Other than those, I think the PR is looking really good

**Review by trgiangdo — COMMENTED** [review]
You also need to remove the `_OPTIONAL_AWS_S3_OBJECT_PARAMETERS_PROPERTY` in line 263 of the `data_node_config.py` file, and add the new _OPTIONAL_.. constant to the `_STORAGE_TYPE_VALUE_S3_OBJECT` dict (as pointed out by the linter)

**Comment by trgiangdo** [comment]
There are still some trailing spaces left in the code, as suggested by the linter [here](https://github.com/Avaiga/taipy/actions/runs/11139315433/job/30957025142?pr=1868).

Other than that, I believe the PR is good to go, when all tests pass

**Review by trgiangdo — DISMISSED** [review]
Hello @AdeshGhadage, thank you again for your contribution. The PR is ready to be merged.

However, due to the fact that Taipy is in the middle of a release, we will have to freeze all code changes for a few days.

In the meantime, feel free to take another issue to work on. We will merge this PR when the release is completed.

File: taipy/core/config/data_node_config.py
Comment (jrobinAV) [inline taipy/core/config/data_node_config.py]:
```suggestion
            aws_s3_object_key (str): The Amazon Web Services (AWS) S3 object key to read
                or write.
```

File: taipy/core/config/data_node_config.py
Comment (jrobinAV) [inline taipy/core/config/data_node_config.py]:
```suggestion
            aws_access_key (str): Amazon Web Services (AWS) ID for to identify account.
            aws_secret_access_key (str): Amazon Web Services (AWS) access key to authenticate
                programmatic requests.
            aws_s3_bucket_name (str): The Amazon Web Services (AWS) S3 bucket to read from and 
                to write the data to.
```

File: taipy/core/config/data_node_config.py
Comment (jrobinAV) [inline taipy/core/config/data_node_config.py]:
```suggestion
            aws_s3_client_parameters (Optional[dict]): Additional parameters for advanced use 
                cases to be passed to the Amazon Web Services (AWS) S3 client.<br/>
                Each parameter key must match the name of a parameter of the
                `boto3.session.Session.client` API.
```

File: taipy/core/config/data_node_config.py
Comment (jrobinAV) [inline taipy/core/config/data_node_config.py]:
```suggestion
            aws_s3_get_object_parameters (Optional[dict]): Additional parameters to be 
                passed to the Amazon Web Services (AWS) S3 client get function for
                advanced reading use cases. <br/> 
                Each parameter key must match the name of a parameter of the
                `boto3.client.get_object` API.
```

File: taipy/core/config/data_node_config.py
Comment (jrobinAV) [inline taipy/core/config/data_node_config.py]:
```suggestion
            aws_s3_put_object_parameters (Optional[dict]): Additional parameters to be 
                passed to the Amazon Web Services (AWS) S3 client put function for
                advanced writing use cases. <br/> 
                Each parameter key must match the name of a parameter of the
                `boto3.client.put_object` API.
```

**Review by jrobinAV — CHANGES_REQUESTED** [review]

**Review by trgiangdo — DISMISSED** [review]
There's currently a conflict change with develop branch. Please resolve it so we can be ready for the merge

**Comment by AdeshGhadage** [comment]
> There's currently a conflict change with develop branch. Please resolve it so we can be ready for the merge

done

File: taipy/core/data/aws_s3.py
Comment (jrobinAV) [inline taipy/core/data/aws_s3.py]: We don't document the Data node attributes like that anymore. We add a new docstring for each public attribute right after its declaration. In this class, None of the attributes are directly exposed (only the ones from the parent classes), so there is nothing to add at the attribute level.

However, we can describe what must/can contain the properties in the dictionary as follows.

```suggestion
    The *properties* attribute must contain the following required entries:
    
    - *aws_access_key* (`str`): Amazon Web Services (AWS) ID to identify the account.
    - *aws_secret_access_key* (`str`): Amazon Web Services (AWS) access key to 
        authenticate programmatic requests.
    - *aws_s3_bucket_name* (`str`): The Amazon Web Services (AWS) S3 bucket to read 
        from and to write the data to.
    - *aws_s3_object_key* (`str`): The Amazon Web Services (AWS) S3 object key to read
        or write.
       
    The *properties* attribute can also contain the following optional entries:
    
    - *aws_region* (`Optional[str]`): Self-contained geographic area where Amazon Web 
        Services (AWS) infrastructure is located.
    - *aws_s3_object_parameters* (`Optional[dict[str, any]]`): A dictionary of 
        additional arguments to be passed into the AWS S3 bucket access string.
    - *aws_s3_client_parameters* (`Optional[dict]`): Additional parameters for advanced 
        use cases to be passed to the Amazon Web Services (AWS) S3 client.<br/>
        Each parameter key must match the name of a parameter of the
        `boto3.session.Session.client` API.
    - *aws_s3_get_object_parameters* (`Optional[dict]`): Additional parameters to be
        passed to the Amazon Web Services (AWS) S3 client get function for
        advanced reading use cases. <br/>
        Each parameter key must match the name of a parameter of the 
        `boto3.client.get_object` API.
    - *aws_s3_put_object_parameters* (`Optional[dict]`): Additional parameters to be
        passed to the Amazon Web Services (AWS) S3 client put function for advanced
        writing use cases. <br/>
        Each parameter key must match the name of a parameter of the
        `boto3.client.put_object` API.
```

**Review by jrobinAV — COMMENTED** [review]
Resolve merging conflicts

**Review by jrobinAV — DISMISSED** [review]

File: taipy/core/data/aws_s3.py
Comment (jrobinAV) [inline taipy/core/data/aws_s3.py]:
```suggestion
    - *aws_secret_access_key* (`str`): Amazon Web Services (AWS) access key to
```

File: taipy/core/data/aws_s3.py
Comment (jrobinAV) [inline taipy/core/data/aws_s3.py]:
```suggestion
    
    - *aws_region* (`Optional[str]`): Self-contained geographic area where Amazon Web
        Services (AWS) infrastructure is located.
    - *aws_s3_object_parameters* (`Optional[dict[str, any]]`): A dictionary of
        additional arguments to be passed into the AWS S3 bucket access string.
    - *aws_s3_client_parameters* (`Optional[dict]`): Additional parameters for advanced
        use cases to be passed to the Amazon Web Services (AWS) S3 client.<br/>
        Each parameter key must match the name of a parameter of the
        `boto3.session.Session.client` API.
    - *aws_s3_get_object_parameters* (`Optional[dict]`): Additional parameters to be
        passed to the Amazon Web Services (AWS) S3 client get function for
        advanced reading use cases. <br/>
        Each parameter key must match the name of a parameter of the
        `boto3.client.get_object` API.
    - *aws_s3_put_object_parameters* (`Optional[dict]`): Additional parameters to be
        passed to the Amazon Web Services (AWS) S3 client put function for advanced
        writing use cases. <br/>
        Each parameter key must match the name of a parameter of the
        `boto3.client.put_object` API.
```

File: taipy/core/data/aws_s3.py
Comment (jrobinAV) [inline taipy/core/data/aws_s3.py]:
```suggestion
    - *aws_s3_bucket_name* (`str`): The Amazon Web Services (AWS) S3 bucket to read
        from and to write the data to.
    - *aws_s3_object_key* (`str`): The Amazon Web Services (AWS) S3 object key to read
        or write.

```

File: taipy/core/data/aws_s3.py
Comment (jrobinAV) [inline taipy/core/data/aws_s3.py]:
```suggestion

```

**Review by jrobinAV — COMMENTED** [review]
ruff checks

File: taipy/core/data/aws_s3.py
Comment (jrobinAV) [inline taipy/core/data/aws_s3.py]:
```suggestion

```

**Comment by jrobinAV** [comment]
@AdeshGhadage I fixed the merge on the docstring as we changed how we document data node attributes. 

FYI, the Update config.pyi and coverage actions are failing because of unauthorized write access issues on our side. Not a problem for you.

**Review by jrobinAV — APPROVED** [review]

**Comment by RymMichaut** [comment]
Congratulations, @AdeshGhadage, for this merged Hacktoberfest PR.
Can you please contact me at rym.michaut@taipy.io to prepare your swag and send it to you ASAP.

**Pull request merged**
